### PR TITLE
RP2040 emulated EEPROM.

### PR DESCRIPTION
**Pull request by tzarc**
## Description

As it says on the title.
Docs forthcoming.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
Having the EEPROM emulation so fast after the RP2040 merge to develop is really awesome! Thanks for your hard work :tada: I have tested the code and it works fine on a regular RPI Pico, Sparkfun RP2040, Elite-pi and Splinky boards - all with different QSPI flash chips.

Reading through the code I noticed some opportunities for clean-up and improvements that are due to the fact that we know the exact use-cases for the flashing code before hand. I'm not sure how much time the bulk write saves but attaching and detaching the whole qspi flash for a single 16bit write seems a bit excessive. I'll attached my findings as patch file as a **suggestion** how I would solve it.

[rp2040_wear_level.patch.txt](https://github.com/qmk/qmk_firmware/files/9030923/rp2040_wear_level.patch.txt)

**Comment by tzarc** [comment]
@KarlK90 -- added, tweaked to do bulk actual read/write operations instead of piecemeal. Seems to work fine.

File: platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            bulk_write_buffer[i] = ~(*values);
+        }
+        __compiler_memory_barrier();
+
+        connect_internal_flash();
+        flash_exit_xip();
+        flash_enable_write();
+
+        flash_put_cmd_addr(FLASHCMD_PAGE_PROGRAM, flash_address);
+        flash_put_get((uint8_t *)bulk_write_buffer, NULL, batch_size * sizeof(backing_store_int_t), 4);
```
Comment (KarlK90) [inline platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c:148]: Ah yes! That is even better :tada:

File: platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash_config.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#endif // WEAR_LEVELING_BACKING_SIZE
+
+// 32kB logical EEPROM
+#ifndef WEAR_LEVELING_LOGICAL_SIZE
+#    define WEAR_LEVELING_LOGICAL_SIZE 4096
+#endif // WEAR_LEVELING_LOGICAL_SIZE
+
+// Define how much flash space we have (defaults to lib/pico-sdk/src/boards/include/boards/***)
+#ifndef WEAR_LEVELING_RP2040_FLASH_SIZE
+#    define WEAR_LEVELING_RP2040_FLASH_SIZE (PICO_FLASH_SIZE_BYTES)
```
Comment (KarlK90) [inline platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash_config.h:26]: I guess we should revisit this once we have the information in the QMK repository.

**Review by KarlK90 — APPROVED** [review]
I have tested the changes with all the boards that I have at hand and it works just fine. When testing it with the wear leveling "test keyboard"  the scan rate improved from 9k to 11k, that is nice! Should `rp2040_flash` be added to the wear leveling docs? Otherwise this is good to go in from my side.

File: platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        _flash_do_cmd(FLASHCMD_READ_STATUS, NULL, &stat, 1);
+    } while (stat & 0x1 && !flash_was_aborted());
+}
+
+// Set the WEL bit (needed before any program/erase operation)
+static void __no_inline_not_in_flash_func(flash_enable_write)(void) {
+    _flash_do_cmd(FLASHCMD_WRITE_ENABLE, NULL, NULL, 0);
+}
+
+static void __no_inline_not_in_flash_func(pico_program_bulk)(uint32_t flash_address, uint16_t *values, size_t item_count) {
```
Comment (KarlK90) [inline platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c:128]:
```suggestion
static void __no_inline_not_in_flash_func(pico_program_bulk)(uint32_t flash_address, backing_store_int_t *values, size_t item_count) {
```

File: platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    rom_connect_internal_flash_fn connect_internal_flash = (rom_connect_internal_flash_fn)rom_func_lookup_inline(ROM_FUNC_CONNECT_INTERNAL_FLASH);
+    rom_flash_exit_xip_fn         flash_exit_xip         = (rom_flash_exit_xip_fn)rom_func_lookup_inline(ROM_FUNC_FLASH_EXIT_XIP);
+    rom_flash_flush_cache_fn      flash_flush_cache      = (rom_flash_flush_cache_fn)rom_func_lookup_inline(ROM_FUNC_FLASH_FLUSH_CACHE);
+    assert(connect_internal_flash && flash_exit_xip && flash_flush_cache);
+
+    static uint16_t bulk_write_buffer[WEAR_LEVELING_RP2040_FLASH_BULK_COUNT];
```
Comment (KarlK90) [inline platforms/chibios/drivers/wear_leveling/wear_leveling_rp2040_flash.c:134]:
```suggestion
    static backing_store_int_t bulk_write_buffer[WEAR_LEVELING_RP2040_FLASH_BULK_COUNT];
```

**Review by KarlK90 — APPROVED** [review]
Nice, LGTM

**Comment by 0xcharly** [comment]
A quick test on the Splinky (RP2040-based board, https://github.com/plut0nium/0xB2), and I can confirm that Via now works out of the box.

Thanks for the continuous efforts to support RP2040! 🙏

**Comment by myst729** [comment]
> A quick test on the Splinky (RP2040-based board, https://github.com/plut0nium/0xB2), and I can confirm that Via now works out of the box.
> 
> Thanks for the continuous efforts to support RP2040! 🙏

@0xcharly  Could you share your sample code? I can build the firmware successfully now but VIA doesn't work with it.

**Comment by 0xcharly** [comment]
> @0xcharly  Could you share your sample code? I can build the firmware successfully now but VIA doesn't work with it.

You can have a look at any of the `bastardkb/**/splinky`. You still need to side load the descriptor file (JSON) first for it to work.

**Comment by myst729** [comment]
@0xcharly Thanks, it works with VIA v2 now. However, VIA v2 seems buggy, compared with v1.

**Pull request merged**